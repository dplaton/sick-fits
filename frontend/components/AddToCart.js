import React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";

import Error from './ErrorMessage';

const ADD_TO_CART_MUTATION = gql`
    mutation addToCart($id: ID!) {
        addToCart(id: $id) {
            id
            quantity
        }
    }
`;

class AddToCart extends React.Component {
    render() {
        const { id } = this.props;

        return (
            <Mutation mutation={ADD_TO_CART_MUTATION} variables={{ id }}>
                {(addToCart, {error})=>(
                    <button onClick={addToCart}>Add to cart ðŸ›’</button>
                )}
            </Mutation>
        );
    }
}

export default AddToCart;
