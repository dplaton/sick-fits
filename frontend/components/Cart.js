import React from 'react';
import CartStyles from './styles/CartStyles';
import SickButton from './styles/SickButton';
import CloseButton from './styles/CloseButton';
import Supreme from './styles/Supreme';

const Cart = () => {
    return(
        <CartStyles open>
            <header>
                <CloseButton title="close">&times;</CloseButton>
                <Supreme>Your shopping cart</Supreme>
                <p>You have __ items in cart</p>
            </header>
            <footer>
                <p>$10.10</p>
                <SickButton>Checkout</SickButton>
            </footer>
        </CartStyles>
    )
}

export default Cart;