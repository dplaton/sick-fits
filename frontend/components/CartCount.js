import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { TransitionGroup, CSSTransition } from "react-transition-group";

/*
 * We customize the classes of the component so we have a flipping effect   
*/
const AnimationStyles = styled.span`
    position: relative;
    .count {
        display: block;
        position: relative;
        transition: all .4s;
        backface-visibility: hidden;
    }
    .count-enter {
        transform: scale(4) rotateX(0.5turn);
    }
    .count-enter-active {
        transform: rotateX(0);
    }
    .count-exit {
        top:0;
        position:absolute;
        transform: rotateX(0);
    }
    .count-exit-active {
        transform: scale(4) rotateX(0.5turn);
    }
`;

const Dot = styled.div`
    background: ${props => props.theme.red};
    border-radius: 50%;
    color: white;
    font-weight: 100;
    font-feature-settings: "tnum";
    font-variant-numeric: tabular-nums;
    line-height: 2rem;
    min-width: 3rem;
    margin-left: 1rem;
    padding: 0.5rem;
`;

/*
 * Using the TransitionGroup and the CSSTransition allows us to add a nice animation to the "cart count" badge
 * What React is doing is basically unmounting a component with count = x and mounting another one with count=x+1. 
 * In this proces React exposes a bunch of CSS classes so we can style the components (both the unmounted one and the mounted one)
 * 
 */
const CartCount = ({ count }) => (
    <AnimationStyles>
        <TransitionGroup>
            <CSSTransition
                unmountOnExit
                className="count"
                classNames="count"
                timeout={{ enter: 400, exit: 400 }}
                key={count}
            >
                <Dot>{count}</Dot>
            </CSSTransition>
        </TransitionGroup>
    </AnimationStyles>
);

export default CartCount;
