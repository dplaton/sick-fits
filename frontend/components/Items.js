import React, {Component} from 'react'
import gql from 'graphql-tag'
import {Query} from 'react-apollo'

const ALL_ITEMS_QUERY = gql `
    query ALL_ITEMS_QUERY {
        items {
            id
            title
            price
            description
            image
            largeImage
        }
    }
`;

class Items extends Component {
    render() {
        return (
            <div>
                <Query query={ALL_ITEMS_QUERY}>
                    {
                        ({data, error, loading}) => {
                            if (loading) return <p>Loading...</p>
                            if (error) return <p>Error! {error}</p>

                            return <p>Hey, I found {data.items.length}</p>;
                        }
                    } 
                </Query>
            </div>
        )
    }
}

export default Items