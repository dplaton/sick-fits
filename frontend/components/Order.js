import React, { Component } from "react";
import PropTypes from "prop-types";
import gql from "graphql-tag";
import { Query } from "react-apollo";
import { format } from "date-fns";
import formatMoney from "../utils/formatMoney";

import OrderStyles from "./styles/OrderStyles";
import Error from "./ErrorMessage";
import OrderItem from "./OrderItem";

const SINGLE_ORDER_QUERY = gql`
    query SINGLE_ORDER_QUERY($id: ID!) {
        order(id: $id) {
            id
            charge
            total
            createdAt
            user {
                id
            }
            items {
                id
                title
                description
                image
                quantity
                price
            }
        }
    }
`;

class Order extends Component {
    static propTypes = {
        id: PropTypes.string.isRequired
    };
    render() {
        return (
            <Query query={SINGLE_ORDER_QUERY} variables={{ id: this.props.id }}>
                {({ data, error, loading }) => {
                    if (loading) return <p>Loading...</p>;
                    if (error) return <Error error={error} />;
                    const order = data.order;
                    return (
                        <OrderStyles data-test="order">
                            <p>
                                <span>ID</span>
                                <span>{order.id}</span>
                            </p>
                            <p>
                                <span>Charge</span>
                                <span>{order.charge}</span>
                            </p>
                            <p>
                                <span>Date</span>
                                <span>
                                    {format(
                                        order.createdAt,
                                        "MMMM d, YYYY h:mm a"
                                    )}
                                </span>
                            </p>
                            <p>
                                <span>Total</span>
                                <span>{formatMoney(order.total)}</span>
                            </p>
                            <p>
                                <span>Items count</span>
                                <span>{order.items.length}</span>
                            </p>
                            <div className="items">
                                {data.order.items.map(orderItem => (
                                    <OrderItem
                                        key={orderItem.id}
                                        item={orderItem}
                                    />
                                ))}
                            </div>
                        </OrderStyles>
                    );
                }}
            </Query>
        );
    }
}

export {SINGLE_ORDER_QUERY};
export default Order;
