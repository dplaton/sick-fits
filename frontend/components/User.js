import { Query } from "react-apollo";
import gql from "graphql-tag";
import PropTypes from "prop-types";

const CURRENT_USER_QUERY = gql`
    query {
        me {
            id
            name
            email
            permissions
            cart {
                id
                quantity
                item {
                    id
                    title
                    price
                    image
                    description
                }
            }
        }
    }
`;

// this is how we build our own render prop
const User = props => (
    // the children of this component must be functions which receive the 'payload' as argument
    <Query {...props} query={CURRENT_USER_QUERY}>
        {payload => props.children(payload)}
    </Query>
);

User.propTypes = {
    children: PropTypes.func.isRequired
};

export default User;
export { CURRENT_USER_QUERY };
