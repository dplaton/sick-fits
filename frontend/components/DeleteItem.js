import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';

import { ALL_ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(id: $id) {
            id
        }
    }
`;

class DeleteItem extends Component {
    update = (cache, payload) => {
        // manually update the cache on the client instead of re-fetching everything

        // 1. Read the items from the cache
        const data = cache.readQuery({ query: ALL_ITEMS_QUERY });

        // 2. Filter the deleted item
        data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
        cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
    };

    render() {
        return (
            // relevant docs: https://www.apollographql.com/docs/react/essentials/mutations.html
            // the "udpdate" prop specifies a function to be used as a 'callback' after the mutation was executed
            <Mutation mutation={DELETE_ITEM_MUTATION} variables={{ id: this.props.id }} update={this.update}>
                {(deleteItem, { error }) => (
                    <button
                        onClick={() => {
                            if (confirm('Are you sure you want to delete this item?')) {
                                deleteItem()
                                    .catch(err => { alert(err.message) });
                            }
                        }}>
                        {this.props.children}
                    </button>
                )}
            </Mutation>
        );
    }
}

export default DeleteItem;
