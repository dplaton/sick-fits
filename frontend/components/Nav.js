import Link from "next/link";
import NavStyles from "./styles/NavStyles";
import User from './User'

const Nav = () => (
    <NavStyles>
        <User>
            {({data: {me}}) => {
                // two level destructuring, like payload.data.me
                if (me) {
                    return <p>{me.name}</p>
                } else {
                    return <p>Nobody</p>
                }
            }}
        </User>
        <Link href="/items">
            <a>Shop</a>
        </Link>
        <Link href="/sell">
            <a>Sell</a>
        </Link>
        <Link href="/signup">
            <a>Signup</a>
        </Link>
        <Link href="/orders">
            <a>Orders</a>
        </Link>
        <Link href="/me">
            <a>Account</a>
        </Link>
    </NavStyles>
);

export default Nav;
