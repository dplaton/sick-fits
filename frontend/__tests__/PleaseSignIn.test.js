import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import PleaseSignIn from "../components/PleaseSignIn";
import { CURRENT_USER_QUERY } from "../components/User";

import { fakeUser } from "../utils/testUtils";

const notSignedInMocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: null} }
    }
];

const signedInMocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: fakeUser()} }
    }
]

describe("<PleaseSignIn />", () => {
    const wrapper = mount(
        <MockedProvider mocks={notSignedInMocks}>
            <PleaseSignIn>
                <p>Signed in</p>
            </PleaseSignIn>
        </MockedProvider>
    );
    it("shows the sign in message if the user is not signed in", async () => {

        await wait(0);
        wrapper.update();
        expect(toJson(wrapper.find('Signin'))).toMatchSnapshot();
    });

    it('renders the component if the user is signed in', async () => {
        const SignedIn = () => <p>Signed in!</p>
        const wrapper = mount(
            <MockedProvider mocks={signedInMocks}>
                <PleaseSignIn>
                    <SignedIn/>
                </PleaseSignIn>
            </MockedProvider>
        );

        await wait(0);
        wrapper.update();
        expect(wrapper.contains(<SignedIn/>)).toBe(true);
    })
});
