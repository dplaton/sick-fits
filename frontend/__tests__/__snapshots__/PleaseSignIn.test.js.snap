// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PleaseSignIn /> shows the sign in message if the user is not signed in 1`] = `
<Signin>
  <Mutation
    mutation={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "SIGNIN_MUTATION",
            },
            "operation": "mutation",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "email",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "email",
                        },
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "password",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "password",
                        },
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "signIn",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "email",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "name",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "String",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "email",
                  },
                },
              },
              Object {
                "defaultValue": undefined,
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "String",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "password",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 188,
          "start": 0,
        },
      }
    }
    refetchQueries={
      Array [
        Object {
          "query": Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": undefined,
                "operation": "query",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "me",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "name",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "email",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "permissions",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "orders",
                            },
                            "selectionSet": Object {
                              "kind": "SelectionSet",
                              "selections": Array [
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "id",
                                  },
                                  "selectionSet": undefined,
                                },
                              ],
                            },
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "cart",
                            },
                            "selectionSet": Object {
                              "kind": "SelectionSet",
                              "selections": Array [
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "id",
                                  },
                                  "selectionSet": undefined,
                                },
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "quantity",
                                  },
                                  "selectionSet": undefined,
                                },
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "item",
                                  },
                                  "selectionSet": Object {
                                    "kind": "SelectionSet",
                                    "selections": Array [
                                      Object {
                                        "alias": undefined,
                                        "arguments": Array [],
                                        "directives": Array [],
                                        "kind": "Field",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "id",
                                        },
                                        "selectionSet": undefined,
                                      },
                                      Object {
                                        "alias": undefined,
                                        "arguments": Array [],
                                        "directives": Array [],
                                        "kind": "Field",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "title",
                                        },
                                        "selectionSet": undefined,
                                      },
                                      Object {
                                        "alias": undefined,
                                        "arguments": Array [],
                                        "directives": Array [],
                                        "kind": "Field",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "price",
                                        },
                                        "selectionSet": undefined,
                                      },
                                      Object {
                                        "alias": undefined,
                                        "arguments": Array [],
                                        "directives": Array [],
                                        "kind": "Field",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "image",
                                        },
                                        "selectionSet": undefined,
                                      },
                                      Object {
                                        "alias": undefined,
                                        "arguments": Array [],
                                        "directives": Array [],
                                        "kind": "Field",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "description",
                                        },
                                        "selectionSet": undefined,
                                      },
                                    ],
                                  },
                                },
                              ],
                            },
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 421,
              "start": 0,
            },
          },
        },
      ]
    }
    variables={
      Object {
        "email": "",
        "password": "",
      }
    }
  >
    <Form
      method="POST"
      onSubmit={[Function]}
    >
      <form
        className="Form-sc-1xszr8q-0 krRxky"
        method="POST"
        onSubmit={[Function]}
      >
        <fieldset
          aria-busy={false}
          disabled={false}
        >
          <DisplayError
            error={Object {}}
          />
          <h2>
            Sign in to your account
          </h2>
          <label
            htmlFor="email"
          >
            Email
            <input
              name="email"
              onChange={[Function]}
              placeholder="Your email"
              type="email"
              value=""
            />
          </label>
          <label
            htmlFor="password"
          >
            Password
            <input
              name="password"
              onChange={[Function]}
              placeholder="Your password"
              type="password"
              value=""
            />
          </label>
          <button
            type="submit"
          >
            Signin
          </button>
        </fieldset>
      </form>
    </Form>
  </Mutation>
</Signin>
`;
